subsys/bluetooth/services/dis.c

```diff
@@ -45,105 +48,265 @@ static struct dis_pnp dis_pnp_id = {
+	BT_GATT_CHARACTERISTIC(BT_UUID_DIS_MANUFACTURER_NAME, BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
 			       read_str, NULL, BT_DIS_MANUF_REF),
 #if CONFIG_BT_DIS_PNP
-	BT_GATT_CHARACTERISTIC(BT_UUID_DIS_PNP_ID,
-			       BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
+	BT_GATT_CHARACTERISTIC(BT_UUID_DIS_PNP_ID, BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
 			       read_pnp_id, NULL, &dis_pnp_id),
 #endif
 
 #if defined(CONFIG_BT_DIS_SERIAL_NUMBER)
-	BT_GATT_CHARACTERISTIC(BT_UUID_DIS_SERIAL_NUMBER,
-			       BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
-			       read_str, NULL,
-			       BT_DIS_SERIAL_NUMBER_STR_REF),
+	BT_GATT_CHARACTERISTIC(BT_UUID_DIS_SERIAL_NUMBER, BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
+			       read_str, NULL, BT_DIS_SERIAL_NUMBER_STR_REF),
 #endif
 #if defined(CONFIG_BT_DIS_FW_REV)
-	BT_GATT_CHARACTERISTIC(BT_UUID_DIS_FIRMWARE_REVISION,
-			       BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
+	BT_GATT_CHARACTERISTIC(BT_UUID_DIS_FIRMWARE_REVISION, BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
 			       read_str, NULL, BT_DIS_FW_REV_STR_REF),
 #endif
 #if defined(CONFIG_BT_DIS_HW_REV)
-	BT_GATT_CHARACTERISTIC(BT_UUID_DIS_HARDWARE_REVISION,
-			       BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
+	BT_GATT_CHARACTERISTIC(BT_UUID_DIS_HARDWARE_REVISION, BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
 			       read_str, NULL, BT_DIS_HW_REV_STR_REF),
 #endif
 #if defined(CONFIG_BT_DIS_SW_REV)
-	BT_GATT_CHARACTERISTIC(BT_UUID_DIS_SOFTWARE_REVISION,
-			       BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
+	BT_GATT_CHARACTERISTIC(BT_UUID_DIS_SOFTWARE_REVISION, BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
 			       read_str, NULL, BT_DIS_SW_REV_STR_REF),
 #endif
+#if defined(CONFIG_BT_DIS_UDI)
+	BT_GATT_CHARACTERISTIC(BT_UUID_UDI_FOR_MEDICAL_DEVICES, BT_GATT_CHRC_READ,
+			       BT_GATT_PERM_READ, read_udi, NULL, NULL),
+#endif
+#if defined(CONFIG_BT_DIS_SYSTEM_ID)
+	BT_GATT_CHARACTERISTIC(BT_UUID_DIS_SYSTEM_ID, BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
+			       read_system_id, NULL, NULL),
+#endif
+#if defined(CONFIG_BT_DIS_IEEE_RCDL)
+	BT_GATT_CHARACTERISTIC(BT_UUID_GATT_IEEE_RCDL, BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
+			       read_str, NULL, BT_DIS_IEEE_RCDL_STR_REF),
+#endif
 
 );
 
 #if defined(CONFIG_BT_DIS_SETTINGS)
-static int dis_set(const char *name, size_t len_rd,
-		   settings_read_cb read_cb, void *store)
+#if defined(CONFIG_BT_DIS_UDI)
+#define ADD_NULLTERM(x) (x > 0 ? x + sizeof('\0') : 0)
+static void dis_update_udi_value(const char *new, char *old, settings_read_cb read_cb,
+				 const char *logkey)
+{
+	/*
+	 * The characteristic contains 1 one-byte flag and 4-null-terminated.
+	 * The null-terminators are only present for strings that are in the flags.
+	 */
+	const size_t merged_size = sizeof(uint8_t) +
+				   ADD_NULLTERM(strlen(BT_DIS_UDI_LABEL_STR_REF)) +
+				   ADD_NULLTERM(strlen(BT_DIS_UDI_DI_STR_REF)) +
+				   ADD_NULLTERM(strlen(BT_DIS_UDI_ISSUER_STR_REF)) +
+				   ADD_NULLTERM(strlen(BT_DIS_UDI_AUTHORITY_STR_REF));
```

Comment:
Given that `ADD_NULLTERM` is always used with `strlen` would it make sense to have something like
```suggestion
#define DIS_STR_SIZE(x) ((x[0) != '\0' ? strlen(x) + sizeof('\0') : 0)
static void dis_update_udi_value(const char *new, char *old, settings_read_cb read_cb,
				 const char *logkey)
{
	/*
	 * The characteristic contains 1 one-byte flag and 4-null-terminated.
	 * The null-terminators are only present for strings that are in the flags.
	 */
	const size_t merged_size = sizeof(uint8_t) +
				   DIS_STR_SIZE(BT_DIS_UDI_LABEL_STR_REF) +
				   DIS_STR_SIZE(BT_DIS_UDI_DI_STR_REF) +
				   DIS_STR_SIZE(BT_DIS_UDI_ISSUER_STR_REF) +
				   DIS_STR_SIZE(BT_DIS_UDI_AUTHORITY_STR_REF);
```

Reply:
Fixed two other uses of  `ADD_NULLTERM`